File: backend/app/utils/json_parser.py
import re
import json
import ast
from typing import Any, Optional, Dict, Union


def extract_first_json_object(text: str) -> Optional[str]:
    start_idx = text.find("{")
    if start_idx == -1:
        return None

    depth = 0
    in_string = False
    is_escaped = False

    for i in range(start_idx, len(text)):
        char = text[i]

        if is_escaped:
            is_escaped = False
            continue

        if char == "\\":
            is_escaped = True
            continue

        if char == '"':
            in_string = not in_string
            continue

        if not in_string:
            if char == "{":
                depth += 1
            elif char == "}":
                depth -= 1
                if depth == 0:
                    return text[start_idx : i + 1]

    return None


def extract_first_json_array(text: str) -> Optional[str]:
    start_idx = text.find("[")
    if start_idx == -1:
        return None

    depth = 0
    in_string = False
    is_escaped = False

    for i in range(start_idx, len(text)):
        char = text[i]

        if is_escaped:
            is_escaped = False
            continue

        if char == "\\":
            is_escaped = True
            continue

        if char == '"':
            in_string = not in_string
            continue

        if not in_string:
            if char == "[":
                depth += 1
            elif char == "]":
                depth -= 1
                if depth == 0:
                    return text[start_idx : i + 1]

    return None


def sanitize_json_string(s: str) -> str:
    cleaned = s.strip()
    # Strip markdown code blocks if wrapped
    cleaned = re.sub(r"^```(?:json)?\s*", "", cleaned, flags=re.IGNORECASE)
    cleaned = re.sub(r"\s*```$", "", cleaned, flags=re.IGNORECASE)
    # Strip single-line comments // ...
    cleaned = re.sub(r"//[^\n\r]*", "", cleaned)
    # Strip multi-line comments /* ... */
    cleaned = re.sub(r"/\*[\s\S]*?\*/", "", cleaned)
    # Replace Python literals with valid JSON literals
    cleaned = re.sub(r"\bTrue\b", "true", cleaned)
    cleaned = re.sub(r"\bFalse\b", "false", cleaned)
    cleaned = re.sub(r"\bNone\b", "null", cleaned)
    # Fix invalid escape sequences (e.g. \a, \P, \., etc. that are not valid in JSON)
    cleaned = re.sub(r'\\(?!["\\/bfnrtu]|u[0-9a-fA-F]{4})', r'\\\\', cleaned)
    # Remove trailing commas before } or ]
    cleaned = re.sub(r",\s*([}\]])", r"\1", cleaned)
    return cleaned


def repair_truncated_json(s: str) -> str:
    cleaned = s.strip()
    cleaned = re.sub(r"^```(?:json)?\s*", "", cleaned, flags=re.IGNORECASE)
    cleaned = re.sub(r"\s*```$", "", cleaned, flags=re.IGNORECASE)

    first_brace = cleaned.find("{")
    first_bracket = cleaned.find("[")
    start_idx = 0
    if first_brace != -1 and first_bracket != -1:
        start_idx = min(first_brace, first_bracket)
    elif first_brace != -1:
        start_idx = first_brace
    elif first_bracket != -1:
        start_idx = first_bracket

    cleaned = cleaned[start_idx:]

    in_string = False
    is_escaped = False
    stack = []

    for char in cleaned:
        if is_escaped:
            is_escaped = False
            continue
        if char == "\\":
            is_escaped = True
            continue
        if char == '"':
            in_string = not in_string
            continue
        if not in_string:
            if char in ("{", "["):
                stack.append(char)
            elif char == "}":
                if stack and stack[-1] == "{":
                    stack.pop()
            elif char == "]":
                if stack and stack[-1] == "[":
                    stack.pop()

    if in_string:
        cleaned += '"'

    # Remove incomplete key-value pairs at the end (e.g. , "key": or , "key" or trailing commas)
    cleaned = re.sub(r",\s*$", "", cleaned)
    cleaned = re.sub(r",\s*(\"[^\"]*\"\s*:\s*[^,}\]]*)$", "", cleaned)
    cleaned = re.sub(r",\s*(\"[^\"]*\"\s*:\s*)?$", "", cleaned)
    cleaned = re.sub(r":\s*$", '""', cleaned)

    while stack:
        open_c = stack.pop()
        cleaned = re.sub(r",\s*$", "", cleaned)
        if open_c == "{":
            cleaned += "}"
        elif open_c == "[":
            cleaned += "]"

    return sanitize_json_string(cleaned)


def clean_and_parse_json(text: str) -> Any:
    if not text or not isinstance(text, str):
        raise ValueError("Empty text provided for JSON parsing.")

    raw = text.strip()

    def try_loads(s: str) -> Optional[Any]:
        if not s or not s.strip():
            return None
        # 1. Standard json.loads
        try:
            return json.loads(s)
        except Exception:
            pass
        # 2. Relaxed control character check (strict=False)
        try:
            return json.loads(s, strict=False)
        except Exception:
            pass
        # 3. Sanitized with strict=False
        try:
            sanitized = sanitize_json_string(s)
            return json.loads(sanitized, strict=False)
        except Exception:
            pass
        # 4. Try ast.literal_eval for python-style dicts
        try:
            val = ast.literal_eval(s)
            if isinstance(val, (dict, list)):
                return val
        except Exception:
            pass
        return None

    # 1. Match code fences
    code_block_regex = r"```(?:json)?\s*([\s\S]*?)\s*```"
    matches = re.findall(code_block_regex, raw, flags=re.IGNORECASE)
    for block in matches:
        content = block.strip()
        extracted = extract_first_json_object(content) or extract_first_json_array(content) or content
        res = try_loads(extracted)
        if res is not None:
            return res

    # 2. Extract first matching balanced JSON object or array
    extracted_obj = extract_first_json_object(raw) or extract_first_json_array(raw)
    if extracted_obj:
        res = try_loads(extracted_obj)
        if res is not None:
            return res

    # 3. Direct slice between first { and last } or first [ and last ]
    cleaned = sanitize_json_string(raw)
    first_b = cleaned.find("{")
    last_b = cleaned.rfind("}")
    if first_b != -1 and last_b != -1 and last_b > first_b:
        res = try_loads(cleaned[first_b : last_b + 1])
        if res is not None:
            return res

    first_sq = cleaned.find("[")
    last_sq = cleaned.rfind("]")
    if first_sq != -1 and last_sq != -1 and last_sq > first_sq:
        res = try_loads(cleaned[first_sq : last_sq + 1])
        if res is not None:
            return res

    # 4. Attempt repair of truncated JSON
    try:
        repaired = repair_truncated_json(raw)
        res = try_loads(repaired)
        if res is not None:
            return res
    except Exception:
        pass

    # 5. Fallback: regex extraction of fieldValues if present
    fallback_dict = {}
    fv_match = re.search(r'"fieldValues"\s*:\s*\{([^}]+)\}', raw)
    if fv_match:
        fv_content = fv_match.group(1)
        kvs = re.findall(r'"([^"]+)"\s*:\s*"([^"]*)"', fv_content)
        if kvs:
            fallback_dict["fieldValues"] = {k: v for k, v in kvs}

    if fallback_dict:
        return fallback_dict

    raise ValueError(f"Could not parse valid JSON from LLM response ({len(raw)} chars)")


def normalize_numeric_field_value(val: Any) -> str:
    if val is None:
        return ""
    if isinstance(val, (int, float)):
        return str(val)

    s = str(val).strip()
    if not s:
        return ""

    lower = re.sub(r"[\s_\-–—]+", "", s.lower())
    if lower in ("unknown", "na", "n/a", "notavailable", "none", "null", "nil", "-", "--", "?", "empty", "undefined"):
        return ""

    if ":" in s:
        after_colon = s.split(":")[-1].strip()
        if re.search(r"\d", after_colon):
            s = after_colon

    s = re.sub(r"^[\"'(\[{;:\s]+|[\"')\]};:\s]+$", "", s)
    dot_count = s.count(".")
    s = re.sub(r"[.,:;_\-\s]+$", "", s).strip()

    if not re.search(r"\d", s):
        return ""

    s = re.sub(r"(\d)\s*,\s*(\d)", r"\1\2", s)

    if "+" in s:
        parts = [p.strip() for p in s.split("+")]
        if len(parts) >= 2 and all(re.match(r"^\d+(?:\.\d+)?$", p) for p in parts):
            total = sum(float(p) for p in parts)
            return str(int(total)) if total.is_integer() else str(round(total, 4))

    if dot_count > 1:
        is_neg = s.startswith("-")
        digits = re.sub(r"[^\d]", "", s)
        return f"-{digits}" if is_neg else digits

    if dot_count == 1:
        idx = s.find(".")
        before = re.sub(r"[^\d+-]", "", s[:idx])
        after = re.sub(r"[^\d]", "", s[idx + 1 :])
        if after:
            clean_int = before if before else "0"
            return f"{clean_int}.{after}"
        else:
            digits = re.sub(r"[^\d]", "", s)
            is_neg = s.startswith("-")
            return f"-{digits}" if is_neg else digits

    is_neg = s.startswith("-")
    digits = re.sub(r"[^\d]", "", s)
    return f"-{digits}" if is_neg else digits

File: frontend/lib/utils/jsonParser.ts
/**
 * Utility to reliably extract and parse JSON from LLM responses.
 * Handles markdown code fences, trailing comments/text, balanced braces,
 * trailing commas, and unescaped characters.
 */

/**
 * Finds and extracts the first balanced JSON object {...} from text
 */
export function extractFirstJsonObject(text: string): string | null {
  const startIdx = text.indexOf('{');
  if (startIdx === -1) return null;

  let depth = 0;
  let inString = false;
  let isEscaped = false;

  for (let i = startIdx; i < text.length; i++) {
    const char = text[i];

    if (isEscaped) {
      isEscaped = false;
      continue;
    }

    if (char === '\\') {
      isEscaped = true;
      continue;
    }

    if (char === '"') {
      inString = !inString;
      continue;
    }

    if (!inString) {
      if (char === '{') {
        depth++;
      } else if (char === '}') {
        depth--;
        if (depth === 0) {
          return text.substring(startIdx, i + 1);
        }
      }
    }
  }

  return null;
}

/**
 * Finds and extracts the first balanced JSON array [...] from text
 */
export function extractFirstJsonArray(text: string): string | null {
  const startIdx = text.indexOf('[');
  if (startIdx === -1) return null;

  let depth = 0;
  let inString = false;
  let isEscaped = false;

  for (let i = startIdx; i < text.length; i++) {
    const char = text[i];

    if (isEscaped) {
      isEscaped = false;
      continue;
    }

    if (char === '\\') {
      isEscaped = true;
      continue;
    }

    if (char === '"') {
      inString = !inString;
      continue;
    }

    if (!inString) {
      if (char === '[') {
        depth++;
      } else if (char === ']') {
        depth--;
        if (depth === 0) {
          return text.substring(startIdx, i + 1);
        }
      }
    }
  }

  return null;
}

/**
 * Sanitizes and repairs common LLM JSON formatting defects
 */
export function sanitizeJsonString(str: string): string {
  let cleaned = str.trim();

  // Remove trailing commas before } or ]
  cleaned = cleaned.replace(/,\s*([}\]])/g, '$1');

  // Strip single-line comments // ...
  cleaned = cleaned.replace(/\/\/[^\n\r]*/g, '');

  return cleaned;
}

/**
 * Intelligently repairs truncated JSON strings caused by model output token limits.
 * Balances open strings, trailing incomplete tokens, brackets, and braces.
 */
export function repairTruncatedJson(str: string): string {
  let cleaned = str.trim();

  // Strip leading code fence if present
  cleaned = cleaned.replace(/^```(?:json)?\s*/i, '');
  cleaned = cleaned.replace(/\s*```$/i, '');

  const firstBrace = cleaned.indexOf('{');
  const firstBracket = cleaned.indexOf('[');
  let startIdx = 0;
  if (firstBrace !== -1 && firstBracket !== -1) {
    startIdx = Math.min(firstBrace, firstBracket);
  } else if (firstBrace !== -1) {
    startIdx = firstBrace;
  } else if (firstBracket !== -1) {
    startIdx = firstBracket;
  }
  cleaned = cleaned.substring(startIdx);

  let inString = false;
  let isEscaped = false;
  const stack: string[] = [];

  for (let i = 0; i < cleaned.length; i++) {
    const char = cleaned[i];

    if (isEscaped) {
      isEscaped = false;
      continue;
    }

    if (char === '\\') {
      isEscaped = true;
      continue;
    }

    if (char === '"') {
      inString = !inString;
      continue;
    }

    if (!inString) {
      if (char === '{' || char === '[') {
        stack.push(char);
      } else if (char === '}') {
        if (stack.length > 0 && stack[stack.length - 1] === '{') {
          stack.pop();
        }
      } else if (char === ']') {
        if (stack.length > 0 && stack[stack.length - 1] === '[') {
          stack.pop();
        }
      }
    }
  }

  // If ended inside a string, close it
  if (inString) {
    cleaned += '"';
  }

  // Remove any trailing incomplete key/value or trailing commas
  cleaned = cleaned.replace(/,\s*$/g, '');
  cleaned = cleaned.replace(/,\s*("[^"]*"\s*:\s*[^,}\]]*)$/g, '');
  cleaned = cleaned.replace(/,\s*("[^"]*"\s*:\s*)?$/g, '');

  // Close open brackets/braces in reverse order
  while (stack.length > 0) {
    const open = stack.pop();
    cleaned = cleaned.replace(/,\s*$/g, '');
    if (open === '{') {
      cleaned += '}';
    } else if (open === '[') {
      cleaned += ']';
    }
  }

  return sanitizeJsonString(cleaned);
}

/**
 * Robust JSON parse function that handles markdown wrappers, trailing prose,
 * unbalanced surrounding characters, and malformed syntax.
 */
export function cleanAndParseJson<T = any>(text: string): T {
  if (!text || typeof text !== 'string') {
    throw new Error('Empty text provided for JSON parsing.');
  }

  const raw = text.trim();

  // 1. Try markdown code blocks first (e.g. ```json ... ```)
  const codeBlockRegex = /```(?:json)?\s*([\s\S]*?)\s*```/gi;
  let match: RegExpExecArray | null;
  while ((match = codeBlockRegex.exec(raw)) !== null) {
    if (match[1]) {
      const blockContent = match[1].trim();
      const extracted =
        extractFirstJsonObject(blockContent) ||
        extractFirstJsonArray(blockContent) ||
        blockContent;

      try {
        return JSON.parse(extracted);
      } catch {
        try {
          return JSON.parse(sanitizeJsonString(extracted));
        } catch {
          // Continue searching
        }
      }
    }
  }

  // 2. Extract first matching balanced JSON object or array from raw text
  const extractedObj = extractFirstJsonObject(raw) || extractFirstJsonArray(raw);
  if (extractedObj) {
    try {
      return JSON.parse(extractedObj);
    } catch {
      try {
        return JSON.parse(sanitizeJsonString(extractedObj));
      } catch {
        // Fall through
      }
    }
  }

  // 3. Strip leading/trailing code fences and slice between first { and last }
  let cleaned = raw;
  if (cleaned.startsWith('```')) {
    cleaned = cleaned.replace(/^```(?:json)?\s*/i, '').replace(/\s*```$/i, '');
  }
  const firstBrace = cleaned.indexOf('{');
  const lastBrace = cleaned.lastIndexOf('}');
  if (firstBrace !== -1 && lastBrace !== -1 && lastBrace > firstBrace) {
    cleaned = cleaned.substring(firstBrace, lastBrace + 1);
  }

  try {
    return JSON.parse(cleaned);
  } catch {
    try {
      return JSON.parse(sanitizeJsonString(cleaned));
    } catch {
      // 4. Fallback: repair truncated JSON (handles token limit exhaustion)
      const repaired = repairTruncatedJson(raw);
      return JSON.parse(repaired);
    }
  }
}

/**
 * Universal Numeric Field Normalizer & OCR Repair Helper:
 * Applies standardized numeric normalization rules across all template fields:
 * - Removes spaces between digits (e.g. "9 9 4" -> "994", "1 3 6 3" -> "1363")
 * - Removes unnecessary trailing punctuation such as '.', ',', ':', ';' (e.g. "9 9 4 ." -> "994", "994." -> "994")
 * - Strips OCR formatting artifacts while preserving meaningful decimal points (e.g. "1 2 3 . 0" -> "123.0", "99.4" -> "99.4")
 * - Handles comma thousand separators (e.g. "1,363" -> "1363")
 * - Distinguishes OCR formatting noise from non-numeric text ("unknown", "N/A", "not available" -> "")
 */
export function normalizeNumericFieldValue(val: any): string {
  if (val === null || val === undefined) return '';
  if (typeof val === 'number') {
    return isNaN(val) ? '' : String(val);
  }

  let str = String(val).trim();
  if (!str) return '';

  // 1. Distinguish actual non-numeric text (e.g. "unknown", "N/A", "not available", "none", "null")
  const lower = str.toLowerCase().replace(/[\s_\-–—]+/g, '');
  if (
    lower === 'unknown' ||
    lower === 'na' ||
    lower === 'n/a' ||
    lower === 'notavailable' ||
    lower === 'none' ||
    lower === 'null' ||
    lower === 'nil' ||
    lower === '-' ||
    lower === '--' ||
    lower === '?' ||
    lower === 'empty' ||
    lower === 'undefined'
  ) {
    return '';
  }

  // 2. Strip leading labels or prefixes if present, e.g. "Opening Counter: 9 9 4 ." -> "9 9 4 ."
  if (str.includes(':')) {
    const afterColon = str.split(':').pop()?.trim();
    if (afterColon && /\d/.test(afterColon)) {
      str = afterColon;
    }
  }

  // 3. Strip surrounding quotes, colons, semicolons, brackets
  str = str.replace(/^["'([{;:\s]+|["')\]};:\s]+$/g, '');

  // Count total dots in original raw string before stripping
  const originalDotCount = (str.match(/\./g) || []).length;

  // 4. Strip trailing punctuation such as '.', ',', ':', ';', '-', '_'
  str = str.replace(/[.,:;_\-\s]+$/, '').trim();

  // If no digits remain, return empty
  if (!/\d/.test(str)) {
    return '';
  }

  // 5. Remove thousand separator commas between digits (e.g. "1,363" -> "1363", "1, 363" -> "1363")
  str = str.replace(/(\d)\s*,\s*(\d)/g, '$1$2');

  // 5b. Evaluate mathematical addition expressions (e.g. "300 + 8", "300+8" -> 308, NOT 3008)
  if (str.includes('+')) {
    const addParts = str.split('+').map((p) => p.trim());
    if (addParts.length >= 2 && addParts.every((p) => /^\d+(?:\.\d+)?$/.test(p))) {
      const sum = addParts.reduce((acc, curr) => acc + parseFloat(curr), 0);
      return Number.isInteger(sum) ? String(sum) : String(parseFloat(sum.toFixed(4)));
    }
  }

  // 6. If multiple dots existed in input (e.g. "4. 33 .", "1.2.3."), they are OCR noise artifacts in an integer
  if (originalDotCount > 1) {
    const isNegative = str.startsWith('-');
    const digitsOnly = str.replace(/[^\d]/g, '');
    if (!digitsOnly) return '';
    return isNegative ? `-${digitsOnly}` : digitsOnly;
  }

  // 7. Check for a single valid decimal point vs OCR noise
  const dotMatches = str.match(/\./g);
  const dotCount = dotMatches ? dotMatches.length : 0;

  if (dotCount === 1) {
    const dotIndex = str.indexOf('.');
    const beforeDot = str.substring(0, dotIndex);
    const afterDot = str.substring(dotIndex + 1);

    const intDigits = beforeDot.replace(/[^\d+-]/g, '');
    const decDigits = afterDot.replace(/[^\d]/g, '');

    // Check if dot was immediately followed by a space and multiple integer digits (e.g. "4. 33")
    const isDotSpaceNoise = !beforeDot.endsWith(' ') && afterDot.startsWith(' ') && decDigits.length >= 2;

    if (decDigits.length > 0 && !isDotSpaceNoise) {
      const cleanInt = intDigits.length > 0 ? intDigits : '0';
      return `${cleanInt}.${decDigits}`;
    } else {
      // Dot was trailing or space noise (e.g. "994.", "4. 33")
      const digitsOnly = str.replace(/[^\d]/g, '');
      const isNegative = str.startsWith('-');
      return isNegative ? `-${digitsOnly}` : digitsOnly;
    }
  }

  // Integer with spaced digits or simple number (e.g. "9 9 4", "1 3 6 3", "427")
  const isNegative = str.startsWith('-');
  const digitsOnly = str.replace(/[^\d]/g, '');
  if (!digitsOnly) return '';

  return isNegative ? `-${digitsOnly}` : digitsOnly;
}

export const repairOcrNumericValue = normalizeNumericFieldValue;
